### PR TITLE
Added code to run pg_dump and pg_restore version 14 and above

### PR DESCRIPTION
**Pull request by ShivanshGahlot**
I have searched and replaced all the occurrences of running the pg_dump and pg_restore command in the go project with the path to the version 14 and higher of either.

### CODE REVIEW

File: yb-voyager/src/srcdb/utils.go [hunk 1]
```diff
@@ -26,3 +30,19 @@ func nameContainsCapitalLetter(name string) bool {
 	}
 	return false
 }
+
+func findAllExecutablesInPath(executableName string) ([]string, error) {
+	pathString := os.Getenv("PATH")
+	if pathString == "" {
+		return nil, fmt.Errorf("PATH environment variable is not set")
+	}
+	paths := strings.Split(pathString, ":")
```
Comment (rahulb-yb) [inline yb-voyager/src/srcdb/utils.go:39]: May be better to use the `os.PathListSeparator` here

File: yb-voyager/src/srcdb/postgres.go [hunk 2]
```diff
@@ -180,3 +190,33 @@ func (pg *PostgreSQL) ExportDataPostProcessing(exportDir string, tablesProgressM
...
+
+	for _, path := range paths {
+		cmd := exec.Command(path, "--version")
+		stdout, err := cmd.Output()
+		if err != nil {
+			utils.ErrExit("error in finding version of %v from path %v: %v", executableName, path, err)
+		}
+
+		//example output: pg_restore (PostgreSQL) 14.5
+		version, err := strconv.ParseFloat(regexp.MustCompile(`[ \n]`).Split(string(stdout), -1)[2], 64)
```
Comment (rahulb-yb) [inline yb-voyager/src/srcdb/postgres.go:209]: Minor nit, but this line has low readability, you can consider using a regex match similar to: 
`(?i)pg_[\S]+ \(PostgreSQL\) ([0-9.]+)`
With the help of `pg_regexp.FindStringSubmatch(string(stdout))`, the first group this returns (if at all) would be the version in its entirety.
Feel free to change the regex to be more generic if you see the possibility, I just came up with this from the top of my head

**Review by rahulb-yb — APPROVED** [review]
Some minor nits, otherwise LGTM 👍🏽

File: yb-voyager/cmd/common.go [hunk 3]
```diff
@@ -87,7 +87,9 @@ func getMappingForTableNameVsTableFileName(dataDirPath string) map[string]string
 		time.Sleep(time.Second * 1)
 	}
 
-	pgRestoreCmd := exec.Command("pg_restore", "-l", dataDirPath)
+	var pgRestorePath string
+	srcdb.FindCorrectPGCommandPath(&pgRestorePath, "pg_restore")
```
Comment (amit-yb) [inline yb-voyager/cmd/common.go:91]: Why not just return the path from the function e.g.

    pgRestorePath := srcdb.FindCorrectPGCommandPath("pg_restore")

File: yb-voyager/src/srcdb/postgres.go [hunk 4]
```diff
@@ -4,15 +4,23 @@ import (
...
+	"strings"
 
 	"github.com/jackc/pgx/v4"
 	log "github.com/sirupsen/logrus"
 	"github.com/yugabyte/yb-voyager/yb-voyager/src/datafile"
 	"github.com/yugabyte/yb-voyager/yb-voyager/src/utils"
 )
 
+var PgDumpPath string
+var PgRestorePath string
```
Comment (amit-yb) [inline yb-voyager/src/srcdb/postgres.go:20]: Let's get rid of these two variables. Whenever you need them, just call the Find* function.

File: yb-voyager/src/srcdb/postgres.go [hunk 5]
```diff
@@ -4,15 +4,23 @@ import (
...
+
+const PG_COMMAND_VERSION float64 = 14.0
```
Comment (amit-yb) [inline yb-voyager/src/srcdb/postgres.go:22]: Please, use strings for version comparison. What if there is a `14.2.1` version?

File: yb-voyager/src/srcdb/postgres.go [hunk 6]
```diff
@@ -180,3 +190,33 @@ func (pg *PostgreSQL) ExportDataPostProcessing(exportDir string, tablesProgressM
 	}
 	dfd.Save()
 }
+
+//find correct path of pg command depending on the version required
```
Comment (amit-yb) [inline yb-voyager/src/srcdb/postgres.go:194]: // Given a PG command name ("pg_dump", "pg_restore"), find absolute path of
    // the executable file having version >= `PG_COMMAND_VERSION`.

Always put a white-space after `//` or `/*`.

File: yb-voyager/src/srcdb/postgres.go [hunk 7]
```diff
@@ -180,3 +190,33 @@ func (pg *PostgreSQL) ExportDataPostProcessing(exportDir string, tablesProgressM
...
+func FindCorrectPGCommandPath(pathVariable *string, executableName string) {
```
Comment (amit-yb) [inline yb-voyager/src/srcdb/postgres.go:195]: `GetAbsPathOfPGCommand(cmd string) string`

File: yb-voyager/src/srcdb/postgres.go [hunk 8]
```diff
@@ -180,3 +190,33 @@ func (pg *PostgreSQL) ExportDataPostProcessing(exportDir string, tablesProgressM
...
+	paths, err := findAllExecutablesInPath(executableName)
+	if err != nil {
+		utils.ErrExit("error in finding executables: %v", err)
+	}
+
```
Comment (amit-yb) [inline yb-voyager/src/srcdb/postgres.go:200]: Explicitly add a check for `len(paths) == 0` and show appropriate message that the command is not installed.

File: yb-voyager/src/srcdb/postgres.go (hunk 2 above)
Comment (amit-yb) [inline yb-voyager/src/srcdb/postgres.go:209]: Use string comparison.

    parts := strings.Fields(stdout)
    version = parts[len(parts)-1]

**Review by amit-yb — COMMENTED** [review]
Just a few minor comments.

File: yb-voyager/src/srcdb/postgres.go (hunk 5 above)
Comment (ShivanshGahlot) [reply yb-voyager/src/srcdb/postgres.go:22]: Done

File: yb-voyager/src/srcdb/utils.go (hunk 1 above)
Comment (ShivanshGahlot) [reply yb-voyager/src/srcdb/utils.go:39]: Done

File: yb-voyager/src/srcdb/postgres.go (hunk 2 above)
Comment (ShivanshGahlot) [reply yb-voyager/src/srcdb/postgres.go:209]: I used the code provided by @amit-yb here

File: yb-voyager/src/srcdb/postgres.go [hunk 9]
```diff
@@ -30,7 +35,9 @@ func (pg *PostgreSQL) Connect() error {
 }
 
 func (pg *PostgreSQL) CheckRequiredToolsAreInstalled() {
-	checkTools("pg_dump", "strings", "pg_restore")
+	checkTools("strings")
+	// pgDumpPath, _ := GetAbsPathOfPGCommand("pg_dump")
```
Comment (kneeraj) [inline yb-voyager/src/srcdb/postgres.go:39]: Get rid of the commented lines?

**Review by kneeraj — COMMENTED** [review]
Did you reproduce the issue first locally and then tested it? @ShivanshGahlot

File: yb-voyager/src/srcdb/postgres.go (hunk 9 above)
Comment (ShivanshGahlot) [reply yb-voyager/src/srcdb/postgres.go:38]: Sorry this was an old commit. Just pushed the latest changes.

Comment (ShivanshGahlot) [reply yb-voyager/src/srcdb/postgres.go]: @kneeraj I did test it by adding paths to both pg_dump binary versions 11 and 14 in the PATH. It chooses version 14 and proceeds. I basically replaced the pg_dump in /usr/bin to version 11 and added the absolute path to version 14 in the path variable.

File: yb-voyager/cmd/common.go [hunk 10]
```diff
@@ -87,7 +87,12 @@ func getMappingForTableNameVsTableFileName(dataDirPath string) map[string]string
 		time.Sleep(time.Second * 1)
 	}
 
-	pgRestoreCmd := exec.Command("pg_restore", "-l", dataDirPath)
+	var pgRestorePath string
```
Comment (amit-yb) [inline yb-voyager/cmd/common.go:90]: Not required. The variable will anyway be defined on the next line.

File: yb-voyager/src/srcdb/postgres.go [hunk 11]
```diff
@@ -180,3 +184,41 @@ func (pg *PostgreSQL) ExportDataPostProcessing(exportDir string, tablesProgressM
...
+			return "", err
+		}
+
+		// example output centos: pg_restore (PostgreSQL) 14.5
+		// example output Ubuntu: pg_dump (PostgreSQL) 14.5 (Ubuntu 14.5-1.pgdg22.04+1)
+		currVersion := strings.Fields(string(stdout))[2]
+		if err != nil {
+			err = fmt.Errorf("error in converting version found from string to float: %v", err)
+			return "", err
+		}
```
Comment (amit-yb) [inline yb-voyager/src/srcdb/postgres.go:215]: Not required anymore.

File: yb-voyager/src/srcdb/postgres.go [hunk 12]
```diff
@@ -180,3 +184,41 @@ func (pg *PostgreSQL) ExportDataPostProcessing(exportDir string, tablesProgressM
...
+	if len(paths) == 0 {
+		err = fmt.Errorf("the command %v is not installed", cmd)
+		return "", err
+	}
+
+	for _, path := range paths {
+		cmd := exec.Command(path, "--version")
+		stdout, err := cmd.Output()
+		if err != nil {
+			err = fmt.Errorf("error in finding version of %v from path %v: %v", cmd, path, err)
```
Comment (amit-yb) [inline yb-voyager/src/srcdb/postgres.go:205]: `%w` is the format specifier used to wrap `error` objects.

File: yb-voyager/src/srcdb/postgres.go [hunk 13]
```diff
@@ -180,3 +184,41 @@ func (pg *PostgreSQL) ExportDataPostProcessing(exportDir string, tablesProgressM
...
+			return "", err
+		}
+
+		// example output centos: pg_restore (PostgreSQL) 14.5
+		// example output Ubuntu: pg_dump (PostgreSQL) 14.5 (Ubuntu 14.5-1.pgdg22.04+1)
```
Comment (amit-yb) [inline yb-voyager/src/srcdb/postgres.go:210]: Very good!! 👍

File: yb-voyager/src/srcdb/postgres.go [hunk 14]
```diff
@@ -180,3 +184,41 @@ func (pg *PostgreSQL) ExportDataPostProcessing(exportDir string, tablesProgressM
...
+		currVersion := strings.Fields(string(stdout))[2]
+		if err != nil {
+			err = fmt.Errorf("error in converting version found from string to float: %v", err)
```
Comment (amit-yb) [inline yb-voyager/src/srcdb/postgres.go:213]: `%w` for `err`.

File: yb-voyager/src/srcdb/postgres.go [hunk 15]
```diff
@@ -180,3 +184,41 @@ func (pg *PostgreSQL) ExportDataPostProcessing(exportDir string, tablesProgressM
...
+			return "", err
+		}
+
+		if version.CompareSimple(currVersion, PG_COMMAND_VERSION) > 0 {
```
Comment (amit-yb) [inline yb-voyager/src/srcdb/postgres.go:217]: `>= 0`

**Review by amit-yb — APPROVED** [review]
A few minor nits. Merge the PR after addressing them.

File: yb-voyager/cmd/common.go (hunk 10 above)
Comment (ShivanshGahlot) [reply yb-voyager/cmd/common.go:90]: Yeah. Removed.

File: yb-voyager/src/srcdb/postgres.go (hunk 12 above)
Comment (ShivanshGahlot) [reply yb-voyager/src/srcdb/postgres.go:205]: Oh right

File: yb-voyager/src/srcdb/postgres.go (hunk 14 above)
Comment (ShivanshGahlot) [reply yb-voyager/src/srcdb/postgres.go:213]: Done

File: yb-voyager/src/srcdb/postgres.go (hunk 15 above)
Comment (ShivanshGahlot) [reply yb-voyager/src/srcdb/postgres.go:217]: Yep fixed

**Pull request merged**